### ISSUE TITLE
404 error with latest Nvidia drivers

### ISSUE THREAD

**Issue by oessaid** [nix-community/nixGL#170]
Installation of `nixGLNvidia` fails with latest drivers on Ubuntu 22.04 (LTS):

```
$ NIXPKGS_ALLOW_UNFREE=1 nix profile install github:guibou/nixGL#nixGLNvidia --impure
```

It looks for a driver file that's not available in `https://download.nvidia.com/XFree86/Linux-x86_64/`.

```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:9:12:
            8|
            9|   strict = derivationStrict drvAttrs;
             |            ^
           10|

       … while evaluating derivation 'nixGLNvidia-550.54.15'
         whose name attribute is located at /nix/store/kc7nh9sdr4z8c7xc9j9891v212k2zj8k-source/pkgs/stdenv/generic/make-derivation.nix:270:7

       … while evaluating attribute 'text' of derivation 'nixGLNvidia-550.54.15'
         at /nix/store/kc7nh9sdr4z8c7xc9j9891v212k2zj8k-source/pkgs/build-support/trivial-builders.nix:117:17:
          116|     runCommand name
          117|       { inherit text executable checkPhase meta;
             |                 ^
          118|         passAsFile = [ "text" ];

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: unable to download 'https://download.nvidia.com/XFree86/Linux-x86_64/550.54.15/NVIDIA-Linux-x86_64-550.54.15.run': HTTP error 404

       response body:

       <?xml version="1.0" encoding="iso-8859-1"?>
       <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
                "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
       <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
        <head>
                <title>404 - Not Found</title>
        </head>
        <body>
                <h1>404 - Not Found</h1>
                <script type="text/javascript" src="//obj.ac.bcon.ecdns.net/ec_tpm_bcon.js"></script>
        </body>
       </html>
```

What is the best mitigation for this? I've seen this issue pop up on the repo from time to time (i.e. there were other issues like this one in the past, e.g. [this one](https://github.com/nix-community/nixGL/issues/124))

**Comment by leguteape** [comment]
Yup, happening to me too. The download URL [https://download.nvidia.com/XFree86/Linux-x86_64/550.40.59/NVIDIA-Linux-x86_64-550.40.59.run](https://download.nvidia.com/XFree86/Linux-x86_64/550.40.59/NVIDIA-Linux-x86_64-550.40.59.run) seems broken.

![Screenshot_20240502_140500](https://github.com/nix-community/nixGL/assets/45424904/f9baf7f2-c337-4621-bdff-fa2cd5c6a2be)

The version being downloaded for me is a Vulkan beta driver and requires a different URL altogether. This one ([https://developer.nvidia.com/downloads/vulkan-beta-5504061-linux](https://developer.nvidia.com/downloads/vulkan-beta-5504061-linux)) is an example for the 550.40.61 Vulkan beta driver

**Comment by AkechiShiro** [comment]
Same issue encountered for 535.183.06

**Comment by AkechiShiro** [comment]
In order to fix this issue, would some kind of mirror containing all the Nvidia drivers (the community can also contribute, if any are missing), experimentals/Nvidia Tesla and so on, would be desirable ?

Maybe using `archive.org`, those files can be archived and then a mirror can be setup to also have them ready to be downloaded in an easy way.

**Comment by ulucs** [comment]
@oessaid would installing an available version on the same major+minor with no auto-detection be a possible workaround? As in
```
nix-build -A auto.nixGLNvidia --argstr nvidiaVersion 550.54.14
```

**Comment by Aietes** [comment]
I'm encountering the same issue, but in Rocky Linux. The key problem in my case is that NVIDIA is not providing consistent versions of their drivers on different distribution channels. For Rocky Linux, and all other RHEL distros, the latest version on the download site differs. e.g. at the time of writing, the latest on downloads `.run` is 570.124.**04**, while the `.rpm` version in the RHEL repository is 570.124.**06**:

Driver download: https://download.nvidia.com/XFree86/Linux-x86_64/570.124.04/
RHEL repository: https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/nvidia-driver-libs-570.124.06-1.el9.x86_64.rpm

Using a different version, even if it's a small difference in the minor version, does not work in my experience. One could always manually install a driver, but in the case of RHEL it's way better to get the NVIDIA driver installed and updated via the default package manager and update cycle.

To solve this issue, I've created a [fork](https://github.com/Aietes/nixGL/tree/rhel) with an additional derivation in nixGL, that gets the driver from the RHEL repository from the `.rpm`. This works well for me, and should solve the issue for RHEL users. I've created a pull request #197, it would be nice if it could be merged.

To get the driver from the RHEL `.rpm` instead of the driver download site, simply set `driverSource` to `rhel`. Also ensure that `enable32bits` is changed from its default to `false`, as at this point I did not include the 32bit version of the driver. Here's an example from my `.nix` using `home-manager`:

```nix
{
  nixpkgs = {
    overlays = specialArgs.overlays;
    config = { allowUnfree = true; };
  };

  home.packages = with pkgs; [
    (nixgl.override {
      driverSource = "rhel";
      enable32bits = false;
    }).auto.nixGLDefault
    (config.lib.nixGL.wrap ghostty)
  ];

  nixGL = {
   packages = pkgs.nixgl.override {
    driverSource = "rhel";
    enable32bits = false;
   };
   defaultWrapper = "nvidia";
  };
```

If you face the driver problem and you are on any RHEL distro like Rocky, Fedora, etc., you can use the nixGL overlay from my fork until this is merged:

```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    nixgl = {
      url = "github:Aietes/nixGL/rhel";
    };
  };
  ....
```

Something similar can probably be done for Debian and other distros, using the official NVIDIA driver for a specific distro as the source for nixGL.

**Comment by Aietes** [comment]
Since it seems that the driver version is consistent across Debian, Fedora and RHEL, at least at this point, pulling the driver from the RHEL repository very likely also works on other distros like Fedora or Debian, pending somebody testing this.
